# Randomly drop RNA-Seq reads from CJ's datasets

<!-- content start -->

**Table of Contents**

- [1. Alignment](#1-alignment)
- [2. subSeq](#2-subseq)
- [3. References](#2-references)

<!-- content end -->

## 1. Alignment



## 2. subSeq

* Col0_FeCl3_HK_Day8 vs. Col0_FeCl3_Live_Day8

![Col0_FeCl3_HK_Day8_vs_Col0_FeCl3_Live_Day8](results/Col0_FeCl3_HK_Day8_vs_Col0_FeCl3_Live_Day8.jpg)

* Col0_FeCl3_HK_Day15 vs. Col0_FeCl3_Live_Day15

![Col0_FeCl3_HK_Day15_vs_Col0_FeCl3_Live_Day15](results/Col0_FeCl3_HK_Day15_vs_Col0_FeCl3_Live_Day15.jpg)

* f6h1_FeCl3_HK_Day8 vs. f6h1_FeCl3_Live_Day8

![f6h1_FeCl3_HK_Day8_vs_f6h1_FeCl3_Live_Day8](results/f6h1_FeCl3_HK_Day8_vs_f6h1_FeCl3_Live_Day8.jpg)

* f6h1_FeCl3_HK_Day15 vs. f6h1_FeCl3_Live_Day15

![f6h1_FeCl3_HK_Day15_vs_f6h1_FeCl3_Live_Day15](results/f6h1_FeCl3_HK_Day15_vs_f6h1_FeCl3_Live_Day15.jpg)


## 3. References

1. Robinson DG, Storey JD: **subSeq: determining appropriate sequencing depth through efficient read subsampling.** *Bioinformatics*. 2014;30(23):3424-6.
